src/clustering_quality.jl

```diff
@@ -0,0 +1,314 @@
+ - `dist_matrix::AbstractMatrix`: a ``n×n`` pairwise distance matrix; `dist_matrix[i,j]` is the distance between ``i``-th and ``j``-th points
+
+ # Keyword arguments
+ - `quality_index::Symbol`: quality index to calculate; see below for the supported options
+ - `fuzziness::Real`: clustering fuzziness > 1
+ - `metric::SemiMetric=SqEuclidean()`: `SemiMetric` object that defines the metric/distance/similarity function
+
+When calling `clustering_quality` one can give `centers`, `assignments` or `weights` arguments by hand or provide a single `clustering` argument from which the necessary data will be read automatically.
+
+For clustering without known cluster centers the datapoints are not required, only `dist_matrix` is necessary. If given, `data` and `metric` will be used to calculate distance matrix instead.
+
+# Supported quality indices
+
+Symbols ↑/↓ are quality direction.
+- `:calinski_harabasz`: hard or fuzzy Calinski-Harabsz index (↑) returns the corrected ratio of between cluster centers inertia and within-clusters inertia
+- `:xie_beni`: hard or fuzzy Xie-Beni index (↓) returns ratio betwen inertia within clusters and minimal distance between cluster centers
+- `:davies_bouldin`: Davies-Bouldin index (↓) returns average similarity between each cluster and its most similar one, averaged over all the clusters
+- `:dunn`: Dunn index (↑) returns ratio between minimal distance between clusters and maximal cluster diameter; it does not make use of `centers` argument
+- `:silhouettes`: average silhouette index (↑), for all silhouettes use [`silhouettes`](@ref) method instead; it does not make use of `centers` argument
+Please refer to the [documentation](@ref clustering_quality) for the definitions and usage descriptions of the supported quality indices. 
+
+"""
+function clustering_quality(
+        data::AbstractMatrix{<:Real},           # d×n matrix
+        centers::AbstractMatrix{<:Real},        # d×k matrix
+        assignments::AbstractVector{<:Integer}; # n vector
+        quality_index::Symbol,
+        metric::SemiMetric=SqEuclidean()
+    )
+    d, n = size(data)
+    dc, k = size(centers)
+
+    d == dc || throw(DimensionMismatch("Inconsistent array dimensions for `data` and `centers`."))
+    (1 <= k <= n) || throw(ArgumentError("Number of clusters k must be from 1:n (n=$n), k=$k given."))
+    k >= 2 || throw(ArgumentError("Quality index not defined for the degenerated clustering with a single cluster."))
+    n == k && throw(ArgumentError("Quality index not defined for the degenerated clustering where each data point is its own cluster."))
+    for (i, clu) in enumerate(assignments)
+        (clu in axes(centers, 2)) || throw(ArgumentError("Invalid cluster index: assignments[$i]=$(clu)."))
+    end
+    for i in 1:k
+        i ∉ assignments && @warn "Cluster number $(i) is empty. Clustering quality calculation may not be reliable."
+    end
```

Comment:
Ok, then something like this should be a less expensive check for the empty clusters (just a draft, I have not tried running it):

```suggestion
    seen_clusters = falses(k)
    for (i, clu) in enumerate(assignments)
        (clu in axes(centers, 2)) || throw(ArgumentError("Invalid cluster index: assignments[$i]=$(clu)."))
        seen_clusters[clu] = true
    end
    if !all(seen_clusters)
        empty_clu_ixs = findall(~seen_clusters)
        @warn "Empty cluster(s) detected: $(join(String.(empty_clu_ixs), ", ")). Clustering quality calculation may not be reliable."
    end
```

One problem we have ATM is that this check is not done when you call another variant of `clustering_quality()` method.
So what would be optimal is to add the gaps check to [`check_assignments()` method](https://github.com/JuliaStats/Clustering.jl/blob/83adafa3d3e6853472d56ea2449c4359bfb07d77/src/utils.jl#L80) via the new `empty_clusters = :warn` keyword option,
and them we can just call `check_assignments()` from here and other public methods.

Reply:
Only methods which use cluster centers seem to have a problem with empty clusters, so it might be fine to give warning only in this case.

Reply:
Actually, the better solution would be just to reindex the clusters skipping the empty ones:
```
old2new = cumsum(seen_clusters)
centers = view(centers, :, seen_clusters) # or findall(seen_clusters), whichever is faster, or even centers[:, seen_clusters]
assignments = old2new[assignments]
```
Not much more expensive, and gives the correct results.
And we can keep reporting the empty clusters (without the "incorrect results" notice), because the clusterings with empty clusters don't sound like 100% good quality.

Reply:
So, do you think we should include it only in hard clustering quality interface? And add keyword option as you suggested above?

Reply:
Does fuzzy suffers from empty clusters (zero weight columns) too? If yes, we can do the same for fuzzy
The keyword option referred to the check_assignments() utility function. But here we are implementing more complex logic to workaround empty clusters, not just warn, so in the end we would not be calling check_assignments(). No need to add such option for clustering_quality(), I think - we just warn unconditionally.

Reply:
Fuzzy quality measures are quite robust, as long as weights are non-negative they are completely fine.